### PR TITLE
Remove extra call to deallocate input to move op and throw if input cannot be reallocated 

### PR DESCRIPTION
**Pull request by abhullar-tt**

### CODE REVIEW

**Review by tt-aho — APPROVED** [review]
lgtm.

I am wondering for the no move case, if it's better to assert out or not. Since the user has no control/access to see if they can move a tensor without calling the move op itself, or if the move op is part of a repeated layer, then it might become clunky for the user to try and only have some of the layers have move and the others not to.

**Comment by abhullar-tt** [comment]
> lgtm.
> 
> I am wondering for the no move case, if it's better to assert out or not. Since the user has no control/access to see if they can move a tensor without calling the move op itself, or if the move op is part of a repeated layer, then it might become clunky for the user to try and only have some of the layers have move and the others not to.

Yeah good point, I will add a warning print that the move is a noop for this case and just return. 

I can’t check this PR in yet because bloom for question answering looks like it’s failing due to oom now that the bug of double deallocating is resolved

File: tt_metal/impl/buffers/buffer.cpp [hunk 1]
```diff
@@ -308,7 +308,7 @@ uint64_t Buffer::core_address(uint32_t core_id) const {
 }
 
 void Buffer::deallocate() {
-    if (this->device_ == nullptr or not this->device_->initialized_) {
+    if (this->device_ == nullptr or not this->device_->initialized_ or this->size_ == 0) {
```
Comment (TT-billteng) [inline tt_metal/impl/buffers/buffer.cpp:311]: How does 0 size buffer even come about?

Comment (abhullar-tt) [reply tt_metal/impl/buffers/buffer.cpp:311]: buffer size is set to 0 when it gets deallocated. this is just early exit for buffer that was previously deallocated

**Comment by arakhmati** [comment]
> @arakhmati could you help take a look at `test_performance_of_bloom_for_question_answering` which is failing. It seems like whatever tensor is being moved can actually be deallocated since the test passes with the double deallocate

Sounds good. I will take a look

File: tt_metal/impl/buffers/buffer.cpp (hunk 1 above)
Comment (TT-billteng) [reply tt_metal/impl/buffers/buffer.cpp:311]: if it gets deallocated, is it possible to structure the code such that the Buffer object doesn't exist at all?

Comment (abhullar-tt) [reply tt_metal/impl/buffers/buffer.cpp:311]: it may be, but that might be outside the scope of this PR since its meant to address a P0 bug

**Review by TT-billteng — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]

**Pull request merged**